datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -1046,6 +1059,130 @@ public int cleanUpOrphanConnectorTasks(boolean cleanUpOrphanTasksInConnector) {
+    }
+
+    Set<DatastreamTask> validTasks = getDatastreamTasks();
+    List<String> allConnectors = getAllConnectors();
+
+    int orphanCount = 0;
+    for (String connector : allConnectors) {
+      Map<String, Set<String>> locksByTaskPrefix = getAllConnectorTaskLocks(connector);
+
+      Set<String> validTaskNamesSet = validTasks.stream()
+          .filter(x -> x.getConnectorType().equals(connector))
+          .map(DatastreamTask::getDatastreamTaskName)
+          .collect(Collectors.toSet());
+
+      List<String> orphanLockList = new ArrayList<>();
+
+      locksByTaskPrefix.forEach((taskPrefix, taskList) -> {
+        // if all the task locks in the task prefix needs to be cleaned up, delete the task prefix node as well.
+        AtomicBoolean deleteTaskPrefixNode = new AtomicBoolean(true);
+        taskList.forEach(taskName -> {
+          if (!validTaskNamesSet.contains(taskName)) {
+            orphanLockList.add(KeyBuilder.datastreamTaskLock(_cluster, connector, taskPrefix, taskName));
+          } else {
+            deleteTaskPrefixNode.set(false);
+          }
+        });
+
+        if (deleteTaskPrefixNode.get()) {
+          orphanLockList.add(KeyBuilder.datastreamTaskLockPrefix(_cluster, connector, taskPrefix));
+        }
+      });
+
+      if (orphanLockList.size() > 0) {
+        LOG.warn("Found orphan task locks: {} in connector: {}", orphanLockList, connector);
+        if (cleanUpOrphanTaskLocksInConnector) {
+          _finalOrphanLockList.addAll(orphanLockList);
+        }
+        orphanCount += orphanLockList.size();
+      }
+    }
+
+    if (cleanUpOrphanTaskLocksInConnector) {
+      // waiting for the debounce time to ensure that the task thread should stop processing by then.
+      _orphanLockCleanupFuture = _scheduledExecutorServiceOrphanLockCleanup.schedule(this::cleanUpOrphanLocks,
+          _debounceTimerMs, TimeUnit.MILLISECONDS);
+    }
+    return orphanCount;
+  }
+
+  private void cleanUpOrphanLocks() {
+    _finalOrphanLockList.forEach(t -> {
+      LOG.info("Deleting task lock node {}", t);
+      _zkclient.delete(t);
+    });
+    _finalOrphanLockList.clear();
+  }
+
+  /*
+   * wait for the task lock to release and delete if the lock is owned by dead owner
+   */
+  private void waitForTaskReleaseOrForceIfOwnerIsDead(DatastreamTask task, long timeoutMs, String lockPath) {
+    boolean deadOwner = false;
+    if (_zkclient.exists(lockPath)) {
+      String owner = _zkclient.readData(lockPath, true);
+      if (owner != null && owner.equals(_instanceName)) {
+        LOG.info("{} already owns the lock on {}", _instanceName, task);
+        //release the lock.
+        return;
+      }
+
+      long waitTimeout = timeoutMs;
+      // if the timeout is greater than debounce timer, only then identify and clean the task lock.
+      if (timeoutMs >= _debounceTimerMs) {
+        // check if the owner is dead.
+        if (_liveInstancesProvider != null && !getLiveInstances().contains(owner)) {
```

Comment:
Ideally, this node is never created without owner, its only created and deleted, so owner will never be null. But, even if it is null, it can be marked as dead owner since there is no owner on the lock.

Reply:
Does the getLiveInstances() list type allow nulls? Otherwise contains() throws a NullPointerException.

>     /**
>      * Returns <tt>true</tt> if this list contains the specified element.
>      * More formally, returns <tt>true</tt> if and only if this list contains
>      * at least one element <tt>e</tt> such that
>      * <tt>(o==null&nbsp;?&nbsp;e==null&nbsp;:&nbsp;o.equals(e))</tt>.
>      *
>      * @param o element whose presence in this list is to be tested
>      * @return <tt>true</tt> if this list contains the specified element
>      * @throws ClassCastException if the type of the specified element
>      *         is incompatible with this list
>      * (<a href="Collection.html#optional-restrictions">optional</a>)
>      * @throws NullPointerException if the specified element is null and this
>      *         list does not permit null elements
>      * (<a href="Collection.html#optional-restrictions">optional</a>)
>      */
>     boolean contains(Object o);